File: services/pipeline_service.py
"""Pipeline command preview for v0.5.1 Pipeline API.

Provides preview_pipeline_run() that returns the shell command and metadata
for each pipeline type — without executing anything.
"""

import os

from scripts import config

# Pipeline definitions: type -> (script_name, args, estimate)
_PIPELINES = {
    "daily": {
        "script": "run_daily_pipeline.py",
        "args": "",
        "duration": "~30 seconds (daily mode, single-day simulation)",
        "description": "8-step daily pipeline: ingest → quality → metrics → alerts → AIP → wake → Feishu",
    },
    "full": {
        "script": "run_full_pipeline.py",
        "args": "",
        "duration": "~5 minutes (full mode, all 634 days)",
        "description": "5-step full pipeline: metrics → alerts → AIP → AI decision → Feishu",
    },
    "db_full": {
        "script": "run_db_pipeline.py",
        "args": "--mode full --dimensional",
        "duration": "~2 minutes (DB mode, all data via SQLite)",
        "description": "5-step DB pipeline: init → ingest → metrics → rules → export",
    },
}

_REQUIRED_ENV_VARS = [
    "API_BEARER_TOKEN",
    "FEISHU_APP_ID",
    "FEISHU_APP_SECRET",
    "FEISHU_BASE_APP_TOKEN",
    "FEISHU_CHAT_ID",
]


def _check_env_warnings() -> list[str]:
    """Check for missing env vars or config files that would affect pipeline runs."""
    warnings = []

    # Optional: warn about Feishu env vars
    for var in ["FEISHU_APP_ID", "FEISHU_APP_SECRET", "FEISHU_BASE_APP_TOKEN", "FEISHU_CHAT_ID"]:
        if not os.environ.get(var, ""):
            warnings.append(f"Env var {var} not set — Feishu operations will use dry-run defaults")

    # Warn if LLM key is not set
    if not os.environ.get("LLM_API_KEY", ""):
        warnings.append("LLM_API_KEY not set — AI decision engine will use heuristic fallback")

    # Check config files exist
    if not os.path.exists(config.ALERT_RULES_FILE):
        warnings.append(f"Alert rules config missing: {config.ALERT_RULES_FILE}")
    if not os.path.exists(config.METRICS_FILE):
        warnings.append(f"Metrics config missing: {config.METRICS_FILE}")

    return warnings


def preview_pipeline_run(pipeline_type: str = "daily") -> dict:
    """Preview a pipeline run without executing anything.

    Args:
        pipeline_type: One of 'daily', 'full', 'db_full'.

    Returns:
        dict with keys: command, pipeline_type, estimated_duration,
                        required_env_vars, warnings, description (str).
    """
    if pipeline_type not in _PIPELINES:
        return {
            "command": "",
            "pipeline_type": pipeline_type,
            "estimated_duration": "",
            "required_env_vars": [],
            "warnings": [f"Unknown pipeline type: '{pipeline_type}'. Valid: {', '.join(_PIPELINES.keys())}"],
            "description": "",
        }

    pipe = _PIPELINES[pipeline_type]
    script_path = os.path.join("scripts", pipe["script"])
    command = f"python3 {script_path}"
    if pipe["args"]:
        command += f" {pipe['args']}"

    warnings = _check_env_warnings()

    return {
        "command": command,
        "pipeline_type": pipeline_type,
        "estimated_duration": pipe["duration"],
        "required_env_vars": _REQUIRED_ENV_VARS,
        "warnings": warnings,
        "description": pipe["description"],
    }


def get_available_pipelines() -> list:
    """Return list of available pipeline types with their descriptions."""
    return [
        {"type": t, "description": p["description"]}
        for t, p in _PIPELINES.items()
    ]

File: internal/service/pipeline_service.go
package service

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"baxi/internal/model"
)

// pipelineDefinition holds metadata for a pipeline type.
type pipelineDefinition struct {
	Script      string
	Args        string
	Duration    string
	Description string
}

var pipelineRegistry = map[string]pipelineDefinition{
	"daily": {
		Script:      "run_daily_pipeline.py",
		Args:        "",
		Duration:    "~30 seconds (daily mode, single-day simulation)",
		Description: "8-step daily pipeline: ingest → quality → metrics → alerts → AIP → wake → Feishu",
	},
	"full": {
		Script:      "run_full_pipeline.py",
		Args:        "",
		Duration:    "~5 minutes (full mode, all 634 days)",
		Description: "5-step full pipeline: metrics → alerts → AIP → AI decision → Feishu",
	},
	"db_full": {
		Script:      "run_db_pipeline.py",
		Args:        "--mode full --dimensional",
		Duration:    "~2 minutes (DB mode, all data via SQLite)",
		Description: "5-step DB pipeline: init → ingest → metrics → rules → export",
	},
}

var requiredEnvVars = []string{
	"API_BEARER_TOKEN",
	"FEISHU_APP_ID",
	"FEISHU_APP_SECRET",
	"FEISHU_BASE_APP_TOKEN",
	"FEISHU_CHAT_ID",
}

// PipelineService provides pipeline preview functionality.
type PipelineService struct {
	configDir string
}

// NewPipelineService creates a new PipelineService.
// configDir is the directory containing YAML config files (e.g. "config").
func NewPipelineService(configDir string) *PipelineService {
	if configDir == "" {
		configDir = "config"
	}
	return &PipelineService{configDir: configDir}
}

// PreviewPipelineRun returns a preview of a pipeline run without executing.
func (s *PipelineService) PreviewPipelineRun(pipelineType string) *model.PipelinePreview {
	pipe, ok := pipelineRegistry[pipelineType]
	if !ok {
		validTypes := s.validPipelineTypes()
		return &model.PipelinePreview{
			Command:      "",
			PipelineType: pipelineType,
			Warnings:     []string{fmt.Sprintf("Unknown pipeline type: '%s'. Valid: %s", pipelineType, validTypes)},
		}
	}

	scriptPath := filepath.Join("scripts", pipe.Script)
	command := fmt.Sprintf("python3 %s", scriptPath)
	if pipe.Args != "" {
		command = fmt.Sprintf("%s %s", command, pipe.Args)
	}

	return &model.PipelinePreview{
		Command:           command,
		PipelineType:      pipelineType,
		EstimatedDuration: pipe.Duration,
		RequiredEnvVars:   requiredEnvVars,
		Warnings:          s.checkEnvWarnings(),
		Description:       pipe.Description,
	}
}

// GetAvailablePipelines returns a list of available pipeline types with descriptions.
func (s *PipelineService) GetAvailablePipelines() []model.PipelineInfo {
	types := make([]string, 0, len(pipelineRegistry))
	for t := range pipelineRegistry {
		types = append(types, t)
	}
	sort.Strings(types)

	result := make([]model.PipelineInfo, 0, len(types))
	for _, t := range types {
		result = append(result, model.PipelineInfo{
			Type:        t,
			Description: pipelineRegistry[t].Description,
		})
	}
	return result
}

func (s *PipelineService) validPipelineTypes() string {
	types := make([]string, 0, len(pipelineRegistry))
	for t := range pipelineRegistry {
		types = append(types, t)
	}
	sort.Strings(types)
	var b strings.Builder
	for i, t := range types {
		if i > 0 {
			b.WriteString(", ")
		}
		b.WriteString(t)
	}
	return b.String()
}

func (s *PipelineService) checkEnvWarnings() []string {
	var warnings []string

	for _, v := range []string{"FEISHU_APP_ID", "FEISHU_APP_SECRET", "FEISHU_BASE_APP_TOKEN", "FEISHU_CHAT_ID"} {
		if os.Getenv(v) == "" {
			warnings = append(warnings, fmt.Sprintf("Env var %s not set — Feishu operations will use dry-run defaults", v))
		}
	}

	if os.Getenv("LLM_API_KEY") == "" {
		warnings = append(warnings, "LLM_API_KEY not set — AI decision engine will use heuristic fallback")
	}

	alertRulesFile := filepath.Join(s.configDir, "alert_rules.yml")
	if _, err := os.Stat(alertRulesFile); os.IsNotExist(err) {
		warnings = append(warnings, fmt.Sprintf("Alert rules config missing: %s", alertRulesFile))
	}

	metricsFile := filepath.Join(s.configDir, "metrics.yml")
	if _, err := os.Stat(metricsFile); os.IsNotExist(err) {
		warnings = append(warnings, fmt.Sprintf("Metrics config missing: %s", metricsFile))
	}

	return warnings
}
